// @aws-amplify/backend-storage/lib/factory.js
import * as path from 'path';
import { AmplifyStorage, } from './construct.js';
import { AmplifyUserError } from '@aws-amplify/platform-core';
/**
 * Singleton factory for a Storage bucket that can be used in `resource.ts` files
 */
class AmplifyStorageFactory {
    props;
    importStack;
    generator;
    /**
     * Set the properties that will be used to initialize the bucket
     */
    constructor(props, importStack = new Error().stack) {
        this.props = props;
        this.importStack = importStack;
    }
    /**
     * Get a singleton instance of the Bucket
     */
    getInstance = ({ constructContainer, outputStorageStrategy, importPathVerifier, }) => {
        importPathVerifier?.verify(this.importStack, path.join('amplify', 'storage', 'resource'), 'Amplify Storage must be defined in amplify/storage/resource.ts');
        this.validateName(this.props.name);
        if (!this.generator) {
            this.generator = new AmplifyStorageGenerator(this.props, outputStorageStrategy);
        }
        return constructContainer.getOrCompute(this.generator);
    };
    validateName = (name) => {
        const nameIsAlphanumeric = /^[a-zA-Z0-9]+$/.test(name);
        if (!nameIsAlphanumeric) {
            throw new AmplifyUserError('InvalidResourceNameError', {
                message: `defineStorage name can only contain alphanumeric characters, found ${name}`,
                resolution: 'Change the name parameter of defineStorage to only use alphanumeric characters',
            });
        }
    };
}
class AmplifyStorageGenerator {
    props;
    outputStorageStrategy;
    resourceGroupName = 'storage';
    constructor(props, outputStorageStrategy) {
        this.props = props;
        this.outputStorageStrategy = outputStorageStrategy;
    }
    generateContainerEntry = (scope) => {
        return new AmplifyStorage(scope, `${this.props.name}`, {
            ...this.props,
            outputStorageStrategy: this.outputStorageStrategy,
        });
    };
}
/**
 * Creates a factory that implements ConstructFactory<AmplifyStorage>
 */
export const defineStorage = (props) => new AmplifyStorageFactory(props, new Error().stack);
